### ISSUE TITLE
`GenesisConfig` in a native runtime free world

### ISSUE THREAD

**Issue by bkchr** [paritytech/substrate#13334]
The `GenesisConfig` is currently one of the only things for that we still require the native runtime. It isn't compiled for wasm right now. We also require the native runtime to load a `GenesisConfig` from a JSON file and as the `GenesisConfig` can change with newer native runtimes we got the raw chain spec that only contains the key/value storage pairs of the genesis block. When we remove the native runtime we need a new way to handle the `GenesisConfig`. 

I propose that we put the `GenesisConfig` into the genesis wasm. This means that we start compiling the `GenesisConfig` for wasm, but I would put this behind some rust feature. So, the normal production builds should not include the `GenesisConfig` anymore. This is done to reduce the size of the wasm binary. I would also keep supporting the serialization of the `GenesisConfig` into JSON. If the `GenesisConfig` is part of the genesis wasm, we have some nice advantages. The first one being that we could keep the JSON serialized `GenesisConfig` in the chain spec and even distribute this chain spec as we would call into the genesis wasm to parse and generate the genesis block. This would require that we move out the genesis wasm from the actual genesis field in the chain spec to some dedicated field from where we can load the wasm binary. Another advantage would be that we could finally make the `chain-spec-builder` work properly. Currently this binary isn't really working for chain specs that aren't build from the `kitchensink-runtime`. The `chain-spec-builder` could be passed some crate name that belongs to some runtime and then it would generate the chain spec for this runtime. We could also enable certain features in the runtime etc.

To build what I outlined above, we would need the following:

- Make `GensisConfig` build in wasm. This will require supporting `serde` in the wasm build to serialize/deserialize the `GenesisConfig` to/from JSON.
- Add a new runtime api `GenesisBuilder` (better name should be found) that can be used to build the genesis config, load it etc.
- Change the chain spec to have a dedicated field for the `code`.
- Conditionally compile `GenesisConfig` into the wasm binary. By default the `GenesisConfig` should probably be compiled into the wasm binary. When doing a on-chain build there should exist a feature to disable the `GenesisConfig`. 

Part of: https://github.com/paritytech/substrate/issues/10579

**Comment by michalkucharczyk** [comment]
Required support for serde addressed in: https://github.dev/paritytech/substrate/pull/13027

**Comment by ggwpez** [comment]
>Is anyone really using runtime-formatted JSON specs? All we really need is raw spec which does not require serde.

I think zombienet relies on meddling with the JSON chain spec and other testing tools probably as well.  
Also it is very convenient for debugging…

**Comment by bkchr** [comment]
Yeah and this will open a lot of possibilities. We will be able to ship json chain specs without them breaking, as the wasm file that loads them stays the same. We will be able to write tools that can modify chain specs very easily etc. It improves the user experience quite a lot.

**Comment by xlc** [comment]
I see it is a nice to have not essential. We can always load metadata from wasm and use it to encode/decode the raw values.

If this feature doesn't need to be part of Substrate, maybe it doesn't need to be part of Substrate?

**Comment by bkchr** [comment]
> I see it is a nice to have but not essential. We can always load metadata from wasm and use it to convert between the raw values and JSON. polkadot.js makes this relatively easy.

Good point, if that works good enough we don't need serde/json support.

**Comment by xlc** [comment]
Chopsticks already implemented this feature and it works well.

Custom storage override:
https://github.com/AcalaNetwork/chopsticks/blob/master/configs/kusama.yml

Code that convert yaml/json to hex:
https://github.com/AcalaNetwork/chopsticks/blob/master/packages/chopsticks/src/utils/set-storage.ts

**Comment by bkchr** [comment]
Yeah to hex is easy, the other way around is more the problem. `GenesisConfig` fields are also not directly the storage items and can be from a different layout and then use custom logic to fill the storage. All in all still something that sounds useful to me. Especially as it is really just much easier for people to interact with it. 

We would also then "disable" the `GenesisConfig` for builds that are going on chain as runtime upgrade to remove serde etc. So, there isn't really any downside.

**Comment by michalkucharczyk** [comment]
@bkchr ,
maybe dumb question, but I am not sure what is good decision here:

When drafting GenesisBuilder Runtime API I've put there some functions which work on `json`.
IMO it would be good to return [`serde_json::error`](https://github.com/serde-rs/json/blob/187f7dadc6ab3423128dde5e5598a83617fb18f2/src/error.rs#L175-L189) (`Result` [here](https://github.com/serde-rs/json/blob/187f7dadc6ab3423128dde5e5598a83617fb18f2/src/error.rs#L15-L23)).
But in order to use this type while crossing _runtime border_, it must implement `TypeInfo`.

I see following options:
- do some wrapper for `serde_json::error`, that copies the fields (and assert sizes) and implements the `TypeInfo + From(serde_json::error)`.
- the other approach could be String, but from what I see `format` is rather not used in runtime.

What do you think?

**Comment by xlc** [comment]
> Conditionally compile GenesisConfig into the wasm binary. By default the GenesisConfig should probably be compiled into the wasm binary. When doing a on-chain build there should exist a feature to disable the GenesisConfig

Without this, I don't really want to enable Serialize/Deserialize.

This is a big regression. We only enable Serialize/Deserialize for good reasons and now we are forced to have then included wasm.

**Comment by bkchr** [comment]
But the linker will remove it when not used, so serialize/deserialize will not be part of the wasm. But yes, you need it at compile time.

**Comment by xlc** [comment]
> But the linker will remove it when not used

It is used if the wasm is able to build genesis? So without a feature flag to disable the ability to build genesis, it will be included on the onchain wasm.

**Comment by AurevoirXavier** [comment]
Our parachain v2 is built upon the previous solo chain and parachain. Its genesis state includes all the previous states. Will this make the genesis WASM hundreds of megabytes?

**Comment by xlc** [comment]
@AurevoirXavier the code to generate genesis is in wasm but the genesis state is not in wasm.

**Comment by bkchr** [comment]
> > But the linker will remove it when not used
> 
> It is used if the wasm is able to build genesis? So without a feature flag to disable the ability to build genesis, it will be included on the onchain wasm.

Yes, but this already explained in the issue that the `GenesisConfig` needs to be build conditionally. There needs to be a guide, but this is something that downstream needs to do on its own as we can not control the feature in your runtime. I think up to this day no one besides us uses a `on-chain-build` feature to disable logging for example.

**Comment by michalkucharczyk** [comment]
Support for genesis building in runtime can be implemented as follows (see this PR https://github.com/paritytech/substrate/pull/14310 for full code):
```rust
use frame_support::genesis_builder_helper::{build_config, create_default_config};
sp_api::impl_runtime_apis! {
        ...
        #[cfg(feature = "genesis-builder")]
        impl sp_genesis_builder::GenesisBuilder<Block> for Runtime {
                fn create_default_config() -> Vec<u8> {
                        create_default_config::<RuntimeGenesisConfig>()
                }

                fn build_config(config: Vec<u8>) -> sp_genesis_builder::Result {
                        build_config::<RuntimeGenesisConfig>(config)
                }
        }
```

The implementation that actually uses all `serde` staff can be hidden behind dedicated feature.

When the feature is enabled depends on use-case. As Basti said, for on-chain builds the feature can be disabled and the resulting wasm blob will not contain any serialization code.


Having this approach implemented for polkadot runtime (PR is under construction), the numbers are as follows:

- `release` profile:

|  type | size    | num of funcs |
| ------------- | ------------- | --- |
| with genesis-builder | 7868841 | 906 |
|  no genesis-builder | 8530347 | 0  |

- `release` profile:

| type | size    | num of funcs |
|--------------------|---------|--------------|
|with genesis-builder| 6436520 | 519          |
|  no genesis-builder| 5896270 | 0            |


where `num of funcs` is the number of `*serde*` symbols in wasm binary:
`wasm-objdump -x polkadot_runtime.wasm | grep -e "- func.*serde" | wc -l`

Just a way to check if `serde` was actually optimized by LTO.

As you can see all the `serde`-related symbols were stripped away.

**Comment by kianenigma** [comment]
Once this whole thing is done, please consider reflecting your changes to https://docs.substrate.io/build/genesis-configuration/ cc @wentelteefje @Rashmirreddy

**Comment by bkchr** [comment]
Yes I already talked with @michalkucharczyk to help updating the guides after these changes are merged!
